node.js

```diff
@@ -663,160 +869,239 @@ async function stabilize() {
 }
 
-/* modified 20191022 */
+/**
+ * Directly implements the pseudocode's fix_fingers() method.
+ * 
+ * @version 20191103
+ */
 async function fix_fingers() {
-    /**
-     * Directly implements the pseudocode's fix_fingers() method.
-     */
     // enable debugging output
     const DEBUGGING_LOCAL = false;
+    let n_successor = NULL_NODE;
 
     // i = random index > 1 into finger[]; but really >0 because 0-based
     // random integer within the range (0, m)
     const i = Math.ceil(Math.random() * (HASH_BIT_LENGTH - 1));
     // finger[i].node = find_successor(finger[i].start);
-    FingerTable[i].successor = await find_successor(FingerTable[i].start, _self, _self);
+    try {
+        n_successor = await find_successor(FingerTable[i].start, _self, _self);
+        if (n_successor.id !== null) {
+            FingerTable[i].successor = n_successor;
+        }
+    } catch (err) {
+        // don't change the finger just yet
+    }
     if (DEBUGGING_LOCAL) {
         console.log("\n>>>>>     Fix {", _self.id, "}.FingerTable[", i, "], with start = ", FingerTable[i].start, ".");
         console.log("     FingerTable[", i, "] =", FingerTable[i].successor, "     <<<<<\n");
     }
 }
 
-/* added 20191021 */
+/**
+ * Directly implements the check_predecessor() method from the IEEE version of the paper.
+ * 
+ * @returns : true if predecessor was still reasonable; false otherwise.
+ * @version 20191021
+ */
 async function check_predecessor() {
-    /**
-     * Directly implements the check_predecessor() method from the MIT version of the paper.
-     */
-    if ( (predecessor.id !== null) && (predecessor.id !== _self.id) ) {
+    if ((predecessor.id !== null) && (predecessor.id !== _self.id)) {
         const predecessor_client = caller(`localhost:${predecessor.port}`, PROTO_PATH, "Node");
         try {
             // just ask anything
             const x = await predecessor_client.getPredecessor(_self.id);
         } catch (err) {
             // predecessor = nil;
             predecessor = { id: null, ip: null, port: null };
+            return false;
         }
     }
+    return true;
 }
 
-/* modified 20191021 */
+/**
+ * Checks whether the successor is still responding.
+ * 
+ * This is an original function, not described in either version of the paper - added 20191103.
+ * 
+ * @returns : true if successor was still reasonable; false otherwise.
+ * @version 20191103
+ */
+async function check_successor() {
+    // enable debugging output
+    const DEBUGGING_LOCAL = false;
+
+    if (DEBUGGING_LOCAL) {
+        console.log("{", _self.id, "}.check_successor(", FingerTable[0].successor.id, ")");
+    }
+
+    let n_successor = NULL_NODE;
+    let successor_seems_ok = false;
+    if (FingerTable[0].successor.id == null) {
+        successor_seems_ok = false;
+    } else if (FingerTable[0].successor.id == _self.id) {
+        successor_seems_ok = true;
+    } else {
+        try {
+            // just ask anything
```

Comment:
Maybe instead of having check_successor( )  we can move all or part of the logic to confirm_exist( ) (we can also rename it) that is a more generalize version and I believe it has the same goal (chekcing if a node is alive or not).

Reply:
I have created issue #34 to work on this.